### PR TITLE
Issue 4341 - Fix the Prometheus metrics garbage collector

### PR DESCRIPTION
**Pull request by kmdouglass**
This PR addresses issue #4341 by proposing a fix to the Prometheus metrics garbage collector. As described in the issue, the problem with the GC is that the methods for deleting metrics from the Prometheus client library will fail when the labels that are provided as arguments do not form an exact match. Recently two additional labels were added to the `workerContainers` and `workerVolumes` metrics but the garbage collector logic was not updated to reflect these changes. As a result, the garbage collector now fails to clean up these metrics when a worker disappears.

The solution proposed in this PR is as follows:
- Maintain a map of the labels used for the `workerContainers`, `workerVolumes`, and `workerTasks` metrics as part of the `PrometheusEmitter`'s state.
- The map of labels is nested; the outer map is keyed by the worker name; the inner map is keyed by a serialized form of the metrics labels. The serialization is done because instances of `prometheus.Labels` themselves are not hashable and can therefore not be used in a set-like data structure such as `map[T]struct{}`
- During one cycle of the garbage collector, pass the labels that are associated with old workers to the [GaugeVec.Delete](https://godoc.org/github.com/prometheus/client_golang/prometheus#GaugeVec.Delete) method to remove the old metrics from memory. This avoids having to use the [GaugeVec.DeleteLabelValues](https://godoc.org/github.com/prometheus/client_golang/prometheus#GaugeVec.DeleteLabelValues) method whose input arguments need to be explicitly enumerated.
- The code as it was structured was not easily unit tested for a number of reasons. I created a small dummy interface that allowed unit testing the metric removal logic.

## Testing

- I added two unit tests to verify the logic for deleting old metrics
- I verified that metrics from stale workers are removed in the `docker-compose` environment by exposing port 9090 from the `web` service and setting a prometheus bind port and IP in the `web` command:

```yaml
<snip>
web:                                                                                                                                                                                                             
    build: .                                                                                                                                                                                                       
    image: concourse/concourse:local                                                                                                                                                                               
    volumes:                                                                                                                                                                                                       
    - .:/src                                                                                                                                                                                                       
    command: web --prometheus-bind-port=9090 --prometheus-bind-ip=0.0.0.0                                                                                                                                          
    depends_on: [db]                                                                                                                                                                                               
    ports:                                                                                                                                                                                                         
    - 8080:8080                                                                                                                                                                                                    
    - 9090:9090
<snip>
```

Then I killed the worker service and verified that the metric for the worker disappeared after 5 minutes by `curl`ing the metric from the ATC:

```
watch -n 30 curl -s localhost:9090 | grep ^[^#] | grep concourse_workers
```

### CODE REVIEW

File: atc/metric/emitter/prometheus.go [hunk 1]
```diff
@@ -46,15 +48,32 @@ type PrometheusEmitter struct {
...
+	workerLastSeen         map[string]time.Time
+	mu                     sync.Mutex
 }
 
 type PrometheusConfig struct {
 	BindIP   string `long:"prometheus-bind-ip" description:"IP to listen on to expose Prometheus metrics."`
 	BindPort string `long:"prometheus-bind-port" description:"Port to listen on to expose Prometheus metrics."`
 }
 
+func serializeLabels(labels *prometheus.Labels) string {
```
Comment (marco-m) [inline atc/metric/emitter/prometheus.go]: Needs a comment to explain why we have to do this.

**Comment by marco-m** [comment]
@cirocosta this is ready for review. If you don't have time, could you please suggest somebody from the Concourse team? Thanks! We have some alarms that depend on this metrics and we are getting false positives daily :-)

**Comment by marco-m** [comment]
@pivotal-jamie-klassen we enjoyed reading your review here in the team :-)

I think that one of the root causes of the mess is that Prometheus is a pull system, while Concourse has been designed thinking about push systems (like all other metrics systems except Concourse actually).

I remember various tickets and PRs in Concourse describing this problem of staleness. I pointed @cirocosta to this blog article that opened my eyes: https://www.robustperception.io/its-easy-to-convert-pull-to-push (the author of that blog is one of the Prometheus main committers).

Last but not least, thanks!

**Comment by kmdouglass** [comment]
Indeed @pivotal-jamie-klassen , it was a bit tricky working on this bug, especially because I had heard that you guys were considering refactoring the metrics logic. I hope this PR can help bridge the gap until that happens.

Thanks a lot for the review!

**Comment by cirocosta** [comment]
@marco-m that's a pretty good one! would you mind sharing your thoughts on https://github.com/concourse/concourse/pull/4247? 🤔  thanks!

thanks for the review, @pivotal-jamie-klassen !

**Pull request merged**